sky/spot/controller.py

```diff
@@ -200,7 +133,97 @@ def _handle_signal(self):
+                # of the nodes are preempted.
+                (cluster_status,
+                 handle) = backend_utils.refresh_cluster_status_handle(
+                     spot_controller.cluster_name, force_refresh=True)
+                if cluster_status != global_user_state.ClusterStatus.UP:
+                    # recover the cluster if it is not up.
+                    logger.info(f'Cluster status {cluster_status.value}. '
+                                'Recovering...')
+                    need_recovery = True
+            if not need_recovery:
+                # The job and cluster are healthy, continue to monitor the
+                # job status.
+                continue
+
+        if job_status == job_lib.JobStatus.SUCCEEDED:
+            end_time = spot_utils.get_job_timestamp(
+                spot_controller.backend,
+                spot_controller.cluster_name,
+                get_end_time=True)
+            # The job is done.
+            spot_state.set_succeeded(spot_controller.job_id, end_time=end_time)
+            break
+
+        if job_status == job_lib.JobStatus.FAILED:
+            # Check the status of the spot cluster. If it is not UP,
+            # the cluster is preempted.
+            (cluster_status,
+             handle) = backend_utils.refresh_cluster_status_handle(
+                 spot_controller.cluster_name, force_refresh=True)
+            if cluster_status == global_user_state.ClusterStatus.UP:
+                # The user code has probably crashed.
+                end_time = spot_utils.get_job_timestamp(
+                    spot_controller.backend,
+                    spot_controller.cluster_name,
+                    get_end_time=True)
+                logger.info(
+                    'The user job failed. Please check the logs below.\n'
+                    '== Logs of the user job (ID: '
+                    f'{spot_controller.job_id}) ==\n')
+                spot_controller.backend.tail_logs(
+                    handle, None, spot_job_id=spot_controller.job_id)
+                logger.info(
+                    f'\n== End of logs (ID: {spot_controller.job_id}) ==')
+                spot_state.set_failed(spot_controller.job_id,
+                                      failure_type=spot_state.SpotStatus.FAILED,
+                                      end_time=end_time)
+                break
+        # cluster can be down, INIT or STOPPED, based on the interruption
+        # behavior of the cloud.
+        # Failed to connect to the cluster or the cluster is partially down.
+        # job_status is None or job_status == job_lib.JobStatus.FAILED
+        logger.info('The cluster is preempted.')
+        spot_state.set_recovering(spot_controller.job_id)
+        recovered_time = spot_controller.strategy_executor.recover()
+        spot_state.set_recovered(spot_controller.job_id,
+                                 recovered_time=recovered_time)
+
+
 if __name__ == '__main__':
+    ray.init('auto')
```

Comment:
What do you think about using `ray.init()` here instead of `ray.init('auto')`?

Pros:
* Provides isolation - if something goes wrong in the ray cluster because of this script it won't affect the main ray cluster
* Removes dependency on having ray running, if we ever need to run this controller in isolation in the future.

Cons:
* Added resource cost/overheads

Reply:
Good point! I think the second Pro holds, but for the first point, ray has changed the behavior for `ray.init()` to the following:
> This will autodetect an existing Ray cluster or start a new Ray instance if no existing cluster is found

That is to say using `ray.init()` instead of `ray.init('auto')` will not have the Con and the first Pro mentioned above.